### PR TITLE
controller: reconcile pods on startup

### PR DESCRIPTION
**Pull request by kmabda**
ensure that we don't miss updated made to pods' networks before the controller get started

Signed-off-by: Abdallah Chatila <abdallah@kaloom.com>

**What this PR does / why we need it**:
In case updates were made to pods's networks while the controller is not present on a node we currently miss reconciling them

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer** *(optional)*:
filtering out pods that are of no concern for the controller (e.g. `pod.Spec.HostNetwork == true`) can be ideally be put in a separate function that can be used by both `handlePodUpdate` function on pods updates and `reconcileOnStartup` function included here when doing pods list

### CODE REVIEW

File: pkg/controller/pod.go [hunk 1]
```diff
@@ -131,6 +139,25 @@ func (pnc *PodNetworksController) worker() {
...
+	pods, err := pnc.podsLister.List(labels.Everything())
+	if err != nil {
+		err = fmt.Errorf("failed to list pods on current node: %v", err)
+		return err
+	}
+	for _, pod := range pods {
+		// since there is no such "not has" relation in a field selector,
+		// filter out pods that are of no concern to the controller here
+		if pod.Spec.HostNetwork {
+			continue
```
Comment (maiqueb) [inline pkg/controller/pod.go:152]: would we want to log these ?

File: pkg/controller/pod.go [hunk 2]
```diff
@@ -131,6 +139,25 @@ func (pnc *PodNetworksController) worker() {
 	}
 }
 
+func (pnc *PodNetworksController) reconcileOnStartup() error {
+	pods, err := pnc.podsLister.List(labels.Everything())
```
Comment (maiqueb) [inline pkg/controller/pod.go:143]: shouldn't we limit this to pod's running on the node where this particular instance is running, like done in https://github.com/k8snetworkplumbingwg/multus-dynamic-networks-controller/pull/90/files#diff-723add03c98864a94589c5e1e0dc8c5a47960f3d50173ed24642ba2c12ccff10R134 ?

**Review by maiqueb — COMMENTED** [review]
Some questions.

File: pkg/controller/pod.go (hunk 1 above)
Comment (kmabda) [reply pkg/controller/pod.go:152]: Yeah we could log an entry at debug level for troubleshooting, however a better approach I think would be to delete lines 149-153 and filter out pods with hostnetwork in `processNextWorkItem` function where we can logs a warning and throw an event as a feedback to the user in the case the latter added a network annotation to these pods as suggested in the comment no 2 https://github.com/k8snetworkplumbingwg/multus-dynamic-networks-controller/pull/89#discussion_r1027361540

Will update this PR based on the decision we take on this subject.

File: pkg/controller/pod.go (hunk 2 above)
Comment (maiqueb) [reply pkg/controller/pod.go:143]: Ah of course. Thank you for pointing out that fact. Makes sense.

File: pkg/controller/pod.go (hunk 1 above)
Comment (maiqueb) [reply pkg/controller/pod.go:152]: I understand your point, but, I wonder if it isn't more efficient to filter the events we send to the queue instead.

We could move all this code to a func, and call from both places (here and in the handlePodUpdate).

I'd say something like:
```golang
func ignoreHostNetworkedPods(pod) {
    if pod.Spec.HostNetwork {
        klog.Warningf(
			"rejecting to add interfaces for host networked pod: %s",
			annotations.NamespacedName(newPod.GetNamespace(), newPod.GetName()),
		)
	pnc.Eventf(
	    newPod,
	    corev1.EventTypeWarning, 
	    "InterfaceAddRejected", 
	    rejectInterfaceAddEventFormat(newPod))
        continue
    }
}
```
Does this make sense ? I have no numbers to backup my claim ..

Comment (kmabda) [reply pkg/controller/pod.go:152]: sure, let do this.

Will wait for https://github.com/k8snetworkplumbingwg/multus-dynamic-networks-controller/pull/89 to be merged with the refactored code that uses the new `ignoreHostNetworkedPods` function and update this PR accordingly

**Comment by maiqueb** [comment]
You should rebase now; thanks again for the contribution.

**Comment by kmabda** [comment]
done

File: pkg/controller/pod.go [hunk 3]
```diff
@@ -131,6 +139,39 @@ func (pnc *PodNetworksController) worker() {
...
+		annotations.NamespacedName(pod.GetNamespace(), pod.GetName()),
+	)
+	pnc.Eventf(pod, corev1.EventTypeWarning, "InterfaceAddRejected", rejectInterfaceAddEventFormat(pod))
+}
+
+func (pnc *PodNetworksController) reconcileOnStartup() error {
+	pods, err := pnc.podsLister.List(labels.Everything())
+	if err != nil {
+		err = fmt.Errorf("failed to list pods on current node: %v", err)
+		return err
```
Comment (maiqueb) [inline pkg/controller/pod.go:154]: nit: can we just inline these ? 
```suggestion
		return fmt.Errorf("failed to list pods on current node: %v", err)
```

File: pkg/controller/pod.go [hunk 4]
```diff
@@ -131,6 +139,39 @@ func (pnc *PodNetworksController) worker() {
...
+		// since there is no such "not has" relation in a field selector,
+		// filter out pods that are of no concern to the controller here
+		if pod.Spec.HostNetwork {
+			indexedNetworkSelectionElements := annotations.IndexPodNetworkSelectionElements(pod)
+			if len(indexedNetworkSelectionElements) > 0 {
+				pnc.ignoreHostNetworkedPods(pod)
+			} else {
+				klog.V(logging.Debug).Infof("host networked pod [%s] got filtered out", namespacedName)
+			}
+			continue
```
Comment (maiqueb) [inline pkg/controller/pod.go:167]: nice.

File: cmd/dynamic-networks-controller/networks-controller.go [hunk 5]
```diff
@@ -121,10 +121,18 @@ func listenOnCoLocatedNode() v1coreinformerfactory.SharedInformerOption {
...
 			)
-			options.FieldSelector = fields.OneTermEqualSelector(filterKey, os.Getenv(nodeNameEnvVariable)).String()
+			// The selector for the pods that this controller instance will watch/reconcile
+			selectorSet := fields.Set{
+				// select pods scheduled only on the node on which this controller instance is running
+				"spec.nodeName": os.Getenv(nodeNameEnvVariable),
+				// select pods with a phase Running to avoid interfering with the cni-plugin works
+				// when pods got created/deleted
+				// see https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-phase
+				"status.phase": "Running",
```
Comment (maiqueb) [inline cmd/dynamic-networks-controller/networks-controller.go:133]: nit: we should use `corev1.Running` instead of a string.

**Review by maiqueb — CHANGES_REQUESTED** [review]
I just have a pedantic nit plus a very minor change request ...

I really like how the solution ended up looking like, thanks for it.

File: pkg/controller/pod.go [hunk 6]
```diff
@@ -244,11 +285,7 @@ func (pnc *PodNetworksController) handlePodUpdate(oldObj interface{}, newObj int
...
 
 	if newPod.Spec.HostNetwork {
-		klog.Warningf(
-			"rejecting to add interfaces for host networked pod: %s",
-			annotations.NamespacedName(newPod.GetNamespace(), newPod.GetName()),
-		)
-		pnc.Eventf(newPod, corev1.EventTypeWarning, "InterfaceAddRejected", rejectInterfaceAddEventFormat(newPod))
+		pnc.ignoreHostNetworkedPods(newPod)
 		return
 	}
```
[the remark this answers is not in the document]
Comment (maiqueb) [reply pkg/controller/pod.go:290]: Makes sense to me. 

Would you mind changing that in an isolated commit in this PR ? 

... you may need to adjust the unit tests for it to pass. If you prefer we can tackle that as a follow-up. 
@kmabda

**Review by maiqueb — APPROVED** [review]
Thanks!

**Comment by kmabda** [comment]
done

**Pull request merged**